apps/ejabberd/src/mod_aws_sns.erl

```diff
@@ -0,0 +1,214 @@
+    ?DEBUG("SNS Packet handle~n    from ~p ~n    to ~p~n    packet ~p.", [From, To, Packet]),
+
+    case {get_topic(Host, Packet), xml:get_subtag(Packet, <<"body">>)} of
+        {undefined, _} -> %% Skip if there is no topic set in configuration for the packet type
+            skip;
+        {_, false} -> %% Skip if there is no message body in the packet
+            skip;
+        {Topic, BodyTag} ->
+            FromGUID = user_guid(Host, From),
+            ToGUID = user_guid(Host, To),
+
+            MessageBody = xml:get_tag_cdata(BodyTag),
+            Content = #{from_user_id => FromGUID,
+                        to_user_id => ToGUID,
+                        message => MessageBody},
+
+            TopicARN = make_topic_arn(Host, Topic),
+            Attributes = message_attributes(Host, TopicARN, From, To, message_type(Packet), Packet),
+
+            publish(Host, TopicARN, Content, Attributes)
+    end,
+
+    ok.
+
+%% Handle user_available_hook
+-spec user_present(UserJID :: ejabberd:jid()) -> ok.
+user_present(#jid{} = UserJID) ->
+    user_precence_changed(UserJID, true).
+
+%% Handle unset_presence_hook
+-spec user_not_present(User     :: ejabberd:luser(),
+                       Server   :: ejabberd:lserver(),
+                       Resource :: ejabberd:lresource(),
+                       _Status :: any()) -> ok.
+user_not_present(User, Host, Resource, _Status) ->
+    user_precence_changed(jid:make_noprep(User, Host, Resource), false).
+
+-spec user_precence_changed(UserJID :: ejabberd:jid(), IsOnline :: boolean()) -> ok.
+user_precence_changed(#jid{lserver = Host} = UserJID, IsOnline) ->
+    Topic = opt(Host, presence_updates_topic),
+    case Topic of
+        undefined ->
+            skip;
+        Topic ->
+            UserGUID = user_guid(Host, UserJID),
+            Content = #{user_id => UserGUID, present => IsOnline},
+            TopicARN = make_topic_arn(Host, Topic),
+            Attributes = message_attributes(Host, TopicARN, UserJID, IsOnline),
+            publish(Host, TopicARN, Content, Attributes)
+    end,
+
+    ok.
+
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% @doc Publish notification to AWS SNS service. Content should be valid JSON term
+-spec publish(Host :: ejabberd:lserver(), topic_arn(), Content :: #{}, attributes()) ->
+                     MessageId :: string().
+publish(Host, TopicARN, Content, Attributes) ->
+    EncodedContent = jiffy:encode(Content),
+    erlcloud_sns:publish(topic, TopicARN, EncodedContent, undefined, maps:to_list(Attributes),
+                         aws_handle(Host)).
+
+%% @doc Returns AWS SNS handle base on configured AWS credentials
+-spec aws_handle(Host :: ejabberd:lserver()) -> erlcloud:sns_handle().
+aws_handle(Host) ->
+    AccessKeyId = [_ | _] = opt(Host, access_key_id),
```

Comment:
Do we have to validate it on every push to AWS? Can we move the validation to module's start and crash when the option doesn't validate? This will get feedback to the devops that sth is misconfigured earlier.